src/kernel/pgtbl.c

```diff
@@ -4,184 +4,145 @@
-		for (i = 0; i < NUM_CPU_COS; i++) {
-			printk("cpu %d: flush %llu\n", i, tlb_quiescence[i].last_mandatory_flush);
-		}
-	}
-
-	return quiescent;
+	return chal_pgtbl_cosframe_add(pt, addr, page, flags);
 }
 
+/* This function updates flags of an existing mapping. */
 int
-cap_memactivate(struct captbl *ct, struct cap_pgtbl *pt, capid_t frame_cap, capid_t dest_pt, vaddr_t vaddr)
+pgtbl_mapping_mod(pgtbl_t pt, u32_t addr, u32_t flags, u32_t *prevflags)
 {
-	unsigned long *    pte, cosframe, orig_v;
-	struct cap_header *dest_pt_h;
-	u32_t              flags;
-	int                ret;
+	return chal_pgtbl_mapping_mod(pt, addr, flags, prevflags);
+}
 
-	if (unlikely(pt->lvl || (pt->refcnt_flags & CAP_MEM_FROZEN_FLAG))) return -EINVAL;
+/**
+ * When we remove a mapping, we need to link the vas to a liv_id,
+ * which tracks quiescence for us.
+ */
+int
+pgtbl_mapping_del(pgtbl_t pt, u32_t addr, u32_t liv_id)
+{
+	return chal_pgtbl_mapping_del(pt, addr, liv_id);
+}
 
-	dest_pt_h = captbl_lkup(ct, dest_pt);
-	if (dest_pt_h->type != CAP_PGTBL) return -EINVAL;
-	if (((struct cap_pgtbl *)dest_pt_h)->lvl) return -EINVAL;
+/* 
+ * NOTE: This just removes the mapping. NO liveness tracking! TLB
+ * flush should be taken care of separately (and carefully).
+ */
+int
+pgtbl_mapping_del_direct(pgtbl_t pt, u32_t addr)
+{
+	return chal_pgtbl_mapping_del_direct(pt, addr);
+}
 
-	pte = pgtbl_lkup_pte(pt->pgtbl, frame_cap, &flags);
-	if (!pte) return -EINVAL;
-	orig_v = *pte;
+int
+pgtbl_mapping_scan(struct cap_pgtbl *pt)
+{
+	return chal_pgtbl_mapping_scan(pt);
+}
 
-	if (!(orig_v & PGTBL_COSFRAME) || (orig_v & PGTBL_COSKMEM)) return -EPERM;
+void *
+pgtbl_lkup_lvl(pgtbl_t pt, u32_t addr, u32_t *flags, u32_t start_lvl, u32_t end_lvl)
+{
+	return chal_pgtbl_lkup_lvl(pt, addr, flags, start_lvl, end_lvl);
+}
 
-	assert(!(orig_v & PGTBL_QUIESCENCE));
-	cosframe = orig_v & PGTBL_FRAME_MASK;
+int
+pgtbl_ispresent(u32_t flags)
+{
+	return chal_pgtbl_ispresent(flags);
+}
 
-	ret = pgtbl_mapping_add(((struct cap_pgtbl *)dest_pt_h)->pgtbl, vaddr, cosframe, PGTBL_USER_DEF);
+unsigned long *
+pgtbl_lkup(pgtbl_t pt, u32_t addr, u32_t *flags)
+{
+	return chal_pgtbl_lkup(pt, addr, flags);
+}
 
-	return ret;
+unsigned long *
+pgtbl_lkup_pte(pgtbl_t pt, u32_t addr, u32_t *flags)
+{
+	return chal_pgtbl_lkup_pte(pt, addr, flags);
 }
 
 int
-pgtbl_activate(struct captbl *t, unsigned long cap, unsigned long capin, pgtbl_t pgtbl, u32_t lvl)
+pgtbl_get_cosframe(pgtbl_t pt, vaddr_t frame_addr, paddr_t *cosframe)
 {
-	struct cap_pgtbl *pt;
-	int               ret;
-
-	pt = (struct cap_pgtbl *)__cap_capactivate_pre(t, cap, capin, CAP_PGTBL, &ret);
-	if (unlikely(!pt)) return ret;
-	pt->pgtbl = pgtbl;
+	return chal_pgtbl_get_cosframe(pt, frame_addr, cosframe);
+}
 
-	pt->refcnt_flags = 1;
-	pt->parent       = NULL; /* new cap has no parent. only copied cap has. */
-	pt->lvl          = lvl;
-	__cap_capactivate_post(&pt->h, CAP_PGTBL);
+/* vaddr -> kaddr */
```

Comment:
??

Reply:
yeah, not a very useful comment.

`This returns a kernel address.` would be better, if that is what you mean.